File: common/src/main/java/net/shiroha233/roadweaver/persistence/WorldDataProvider.java
package net.shiroha233.roadweaver.persistence;

import dev.architectury.injectables.annotations.ExpectPlatform;
import net.shiroha233.roadweaver.core.model.StructureConnection;
import net.shiroha233.roadweaver.core.model.StructureInfo;
import net.shiroha233.roadweaver.core.model.StructureLocationData;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * 跨平台世界数据访问抽象，由 Fabric/Forge 平台端提供实现
 */
public abstract class WorldDataProvider {

    @ExpectPlatform
    public static WorldDataProvider getInstance() {
        throw new AssertionError();
    }

    public abstract StructureLocationData getStructureLocations(ServerLevel level);
    public abstract void setStructureLocations(ServerLevel level, StructureLocationData data);

    public abstract List<StructureConnection> getStructureConnections(ServerLevel level);
    public abstract void setStructureConnections(ServerLevel level, List<StructureConnection> connections);

    public abstract List<StructureConnection> getHighwayConnections(ServerLevel level);
    public abstract void setHighwayConnections(ServerLevel level, List<StructureConnection> connections);

    public abstract Set<Long> getPlannedTileKeys(ServerLevel level);
    public abstract void setPlannedTileKeys(ServerLevel level, Set<Long> keys);
    public abstract Map<Long, Long> getPlannedTileCenters(ServerLevel level);
    public abstract void setPlannedTileCenters(ServerLevel level, Map<Long, Long> centers);

    /**
     * 添加单个结构位置（幂等）
     */
    public void addStructureLocation(ServerLevel level, BlockPos pos) {
        StructureLocationData data = getStructureLocations(level);
        List<BlockPos> locations = new ArrayList<>(data != null ? data.structureLocations() : new ArrayList<>());
        List<StructureInfo> infos = new ArrayList<>(data != null ? data.structureInfos() : new ArrayList<>());
        if (!locations.contains(pos)) {
            locations.add(pos);
            setStructureLocations(level, new StructureLocationData(locations, infos));
        }
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/persistence/WorldDataProvider.kt
package net.shiroha233.roadweaver.persistence

import dev.architectury.injectables.annotations.ExpectPlatform
import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.shiroha233.roadweaver.helpers.Records

object WorldDataProviderAccess {
    @JvmStatic
    @ExpectPlatform
    fun getInstance(): WorldDataProvider {
        throw AssertionError()
    }
}

/**
 * 跨平台世界数据访问抽象（Common）。
 * 使用 @ExpectPlatform 提供平台端实现提供者。
 */
abstract class WorldDataProvider {
    companion object {
        @JvmStatic
        fun getInstance(): WorldDataProvider = WorldDataProviderAccess.getInstance()
    }

    // 结构位置（用于幂等性检查）
    abstract fun getStructureLocations(level: ServerLevel): Records.StructureLocationData
    abstract fun setStructureLocations(level: ServerLevel, data: Records.StructureLocationData)

    // 结构连接（道路规划）
    abstract fun getStructureConnections(level: ServerLevel): List<Records.StructureConnection>
    abstract fun setStructureConnections(level: ServerLevel, connections: List<Records.StructureConnection>)

    // 规划覆盖：tile 键集合与中心点映射
    abstract fun getPlannedTileKeys(level: ServerLevel): Set<Long>
    abstract fun setPlannedTileKeys(level: ServerLevel, keys: Set<Long>)
    abstract fun getPlannedTileCenters(level: ServerLevel): Map<Long, Long>
    abstract fun setPlannedTileCenters(level: ServerLevel, centers: Map<Long, Long>)

    // 便捷方法：添加单个结构位置
    fun addStructureLocation(level: ServerLevel, pos: BlockPos) {
        val data = getStructureLocations(level)
        val locations = ArrayList(data.structureLocations)
        if (!locations.contains(pos)) {
            locations.add(pos)
            setStructureLocations(level, Records.StructureLocationData(locations))
        }
    }
}
